gcc/rust/resolve/rust-late-name-resolver-2.0.cc

```diff
@@ -167,9 +220,80 @@ Late::visit (AST::IdentifierPattern &identifier)
   // but values does not allow shadowing... since functions cannot shadow
   // do we insert functions in labels as well?
 
-  // We do want to ignore duplicated data because some situations rely on it.
-  std::ignore = ctx.values.insert_shadowable (identifier.get_ident (),
-					      identifier.get_node_id ());
+  if (ctx.bindings.and_binded (identifier.get_ident ()))
+    {
+      if (ctx.bindings.get_source () == BindingSource::Param)
+	rust_error_at (
+	  identifier.get_locus (), ErrorCode::E0415,
+	  "identifier %qs is bound more than once in the same parameter list",
+	  identifier.as_string ().c_str ());
+      else
+	rust_error_at (
+	  identifier.get_locus (), ErrorCode::E0416,
+	  "identifier %qs is bound more than once in the same pattern",
+	  identifier.as_string ().c_str ());
+      return;
+    }
+
+  ctx.bindings.insert_ident (identifier.get_ident ());
+
+  if (ctx.bindings.or_binded (identifier.get_ident ()))
+    {
+      // FIXME: map usage instead
+      std::ignore = ctx.values.insert_shadowable (identifier.get_ident (),
+						  identifier.get_node_id ());
+    }
+  else
+    {
+      // We do want to ignore duplicated data because some situations rely on
+      // it.
+      std::ignore = ctx.values.insert_shadowable (identifier.get_ident (),
+						  identifier.get_node_id ());
+    }
+}
+
+void
+Late::visit (AST::AltPattern &pattern)
+{
+  ctx.bindings.push (Binding::Kind::Or);
+  for (auto &alt : pattern.get_alts ())
+    {
+      ctx.bindings.push (Binding::Kind::Product);
+      visit (alt);
+      ctx.bindings.merge ();
+    }
+  ctx.bindings.merge ();
+}
+
+void
+Late::visit (AST::Function &function)
+{
+  auto def_fn = [this, &function] () {
+    visit_outer_attrs (function);
+    visit (function.get_visibility ());
+    visit (function.get_qualifiers ());
+    for (auto &generic : function.get_generic_params ())
+      visit (generic);
+
+    // We only care about params
+    ctx.bindings.new_binding (BindingSource::Param);
+
+    for (auto &param : function.get_function_params ())
+      visit (param);
+
+    ctx.bindings.clear ();
```

Comment:
We could introduce some functions in the AST visitor specially dedicated to visiting the parameters, this way we could override those functions instead of the whole visit.